keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,773 @@
+    ......  
+}
+
+type ContainerNotifier struct {
+   // Name must be unique within a container.
+   // Names are label-style, with an optional prefix. Names without the prefix
+   // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+   // Names with a prefix such as `example.com/label-style` are custom names.
+   Name string
+
+   // Action defines what happens when a notifier is triggered.  Exactly-once
+   // execution is not guaranteed, so this must be idempotent. When the action is
+   // triggered multiple times, it should achieve the same results. For example,
+   // when a quiesce command is run on a database, the database is quiesced. If the
+   // same quiesce command is run again, it must not have additional impact,
+   // because the database is already quiesced.
+   Action *ContainerNotifierAction
+
+    // Number of seconds after which the notifier times out.
+    // Defaults to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+}
+
+// ContainerNotifier describes a command to run in the container.
+type ContainerNotifierAction struct {
+    // handler describes how this notifier is delivered to the container.
+    Handler ContainerNotifierHandler
+}
+
+// Handler defines a specific action that should be taken
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+}
+```
+
+##### Notification API Object
+
+The Notification object is a request for a ContainerNotifier. It will be an in-tree API object in the core API group so that Kubelet can monitor and trigger the actions. The external controller or a user creates the Notification object to request a ContainerNotifier.
+
+All containers in a selected pod which have a "ContainerNotifier" defined in their spec with the "ContainerNotifierName" will have their corresponding "ContainerNotifierAction" executed. Moreover, there is no guarantee of ordering on execution.
+
+```
+type Notification struct {
+           metav1.TypeMeta
+           // +optional
+           metav1.ObjectMeta
+
+           // Spec defines the behavior of a notification.
+           // +optional
+           Spec NotificationSpec
+
+           // Status defines the current state of a notification.
+           // +optional
+           Status NotificationStatus
+}
+
+type NotificationSpec struct {
+            // PodSelector specifies a label query over a set of pods.
+            // +optional
+            PodSelector *metav1.LabelSelector
+
+           // This contains the name of the ContainerNotifier to send to a container.
+           // Names are label-style, with an optional prefix. Names without the prefix
+           // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+           // Names with a prefix such as `example.com/label-style` are custom names.
+           ContainerNotifierName string
```

Comment:
minor: This is a needlessly long name.  Just `Notifier` or `NotifierName` perhaps?

Reply:
It would be good to think about who can send notifications ? What permissions they need on a pod ?. Since the controller is doing exec, does it mean as long as exec is opened for the controller, anyone can create and run the commands using controller as a proxy?

Another idea is should we make notifier a subresource ?

Reply:
RBAC governs who can create Notifications in a namepace - isn't that enough?

Reply:
see here https://github.com/kubernetes/enhancements/pull/1995/files#r555640516